### PR TITLE
Fix beam-search requests not terminating at large beam_width

### PR DESCRIPTION
**Pull request by Doloxetine**
## Summary

Under beam search with large `beam_width` (e.g. `beam_width=128`), requests
fail to terminate even after reaching `max_tokens`. Token counts grow
unbounded — from hundreds into the tens of thousands — exhausting KV cache
blocks, blocking the entire in-flight batch, and driving QPS to zero. The
same infinite-generation state subsequently triggers two secondary crashes
(`std::vector > max_size()` and `decoderFinishedEvent must be nullopt`),
causing a permanent failure that affects every subsequent request on the same
micro-batch slot.

This PR fixes the root cause, the two cascading crashes, and one related
generation-logits correctness bug that becomes observable in the same
scenario.

## Root Cause

TRT-LLM currently relies on two termination mechanisms; **both fail under
beam search at large `beam_width`**, because beam reordering is performed
after every decoder step and physically rearranges beam data (KV cache, token
sequences, and finished states).

### 1. GPU-side: `finishedSum == reqBeamWidth`

A beam marked finished at step N can be overwritten by an unfinished beam
at step N+1:

| Step | finishedSum | What happened                                           |
|------|-------------|---------------------------------------------------------|
| 5    | 1           | beam[3] generated EOS                                   |
| 6    | 0           | beam reorder replaced beam[3] with an unfinished beam   |
| 7    | 2           | beam[1], beam[5] generated EOS                          |
| 8    | 1           | beam reorder overwrote some again                       |
| ...  | ...         | never reaches reqBeamWidth                              |

With `beam_width=128`, `finishedSum` oscillates but essentially never
converges to 128.

### 2. CPU-side: `willCompleteNextIteration()`

```cpp
SizeType32 getMaxNumGeneratedTokens() const {
    return getMaxBeamNumTokens() - mPromptLen;
}
bool willCompleteNextIteration() const {
    return getMaxNumGeneratedTokens() + mNumTokensPerIteration >= mMaxNewTokens;
}
```

In `updateRequests()`, new tokens are appended based on
`seqLen - llmReq->getNumTokens(beam)`. After beam reorder,
`sequenceLengths[beam]` may correspond to a **different (shorter) beam**
than the CPU-side `mTokens[beam]` tracked, producing
`numNewOutputTokens <= 0`. As a result `getMaxNumGeneratedTokens()` stalls
and `willCompleteNextIteration()` never returns true.

### Cascading failures

Once termination is forced by any upstream limit:

1. **`cannot create std::vector larger than max_size()`** in
   `postProcessRequest`: `gatherTree`/`finalize` may produce out-of-range
   `sequenceLengths` for still-unfinished beams, so
   `generatedLength = sequenceLengths[beam] - promptLen` becomes
   astronomically large and `std::vector::assign` throws.

2. **`decoderFinishedEvent must be nullopt` assertion** in `forwardAsync`:
   the exception above propagates out of `decoderSync` and skips
   `decoderFinishedEvent.reset()`. The next `forwardAsync` then hits the hard
   assertion, permanently poisoning the micro-batch slot for every
   subsequent request.

### Related correctness bug (same call path)

`generation_logits` are stored indexed by beam **slot** position during
generation. `gatherTree` reorders `output_ids` by tracing `parentIds` to
reconstruct the correct beam paths, but it does **not** reorder
`generation_logits`. As a result the beam dimension of `output_ids` and
`generation_logits` is misaligned whenever any beam reorder happened, and
downstream consumers that correlate the two (e.g. computing per-token
confidence, distillation, rerank) see mismatched values.

### 1. Hard cutoff using `decodingIter` in `updateRequests()`

`decodingIter` is a pure iteration counter incremented unconditionally at
the end of every `updateRequests()`, and is therefore immune to beam
reordering:

```cpp
auto const maxNumGenerated = llmReq->getMaxNumGeneratedTokens();
auto const decodingIter    = llmReq->getDecodingIter();

bool const cutoffByIter    = (decodingIter + 1 >= llmReq->mMaxNewTokens);
bool const cutoffByLength  = (maxNumGenerated    >= llmReq->mMaxNewTokens);
bool const hardMaxTokensCutoff = cutoffByIter || cutoffByLength;

if (hardMaxTokensCutoff && decoderFinishedSumPtr[seqSlot] != reqBeamWidth
    && !llmReq->isGenerationCompleteState())
{
    // Re-derive the reason from the SAME finishReasonsHostData that fed
    // setFinishedReason earlier in this loop so we never clobber a real
    // kEND / kSTOP_WORDS.
    auto const* const beamFinishPtr = finishReasonsHostData + seqSlot * mOperatingBeamWidth;
    for (SizeType32 beam = 0; beam < reqBeamWidth; ++beam)
    {
        if (beamFinishPtr[beam].toFinishReason() == executor::FinishReason::kNOT_FINISHED)
        {
            llmReq->setFinishedReason(executor::FinishReason::kLENGTH, beam);
        }
    }
}
```

`hardMaxTokensCutoff` is then folded into `shouldFinalizeRequest` so the
existing terminate path runs and frees the KV cache.

**Why two conditions?** `cutoffByIter` handles the common beam-reorder case
(`mTokens` not growing on a step); `cutoffByLength` handles speculative
decoding (one iteration may accept multiple tokens).

### 2. Beam-reorder-safe `numNewOutputTokens`

Snapshot the per-beam length at the start of `updateRequests()` and use it
as the baseline, rather than the live `getNumTokens(beam)` which can be
stale for another beam after reordering:

```cpp
std::vector<SizeType32> currentNumTokensPerBeam(reqBeamWidth);
for (SizeType32 beam = 0; beam < reqBeamWidth; ++beam)
{
    currentNumTokensPerBeam[beam] = llmReq->getNumTokens(beam);
}
...
auto const numNewOutputTokens = std::max<SizeType32>(0, seqLen - currentNumTokensPerBeam[beam]);
```

For `beam_width == 1` (sampling/greedy) this is numerically identical to
the previous behavior; the `TLLM_CHECK_WITH_INFO(numGeneratedTokens >=
numNewOutputTokens, ...)` guard is preserved on that path.

### 3. Graceful recovery for `decoderFinishedEvent` in `forwardAsync()`

Replace the hard assertion:

```cpp
// before
TLLM_CHECK_WITH_INFO(!decoderFinishedEvent.has_value(), "decoderFinishedEvent must be nullopt.");

// after
if (decoderFinishedEvent.has_value())
{
    TLLM_LOG_WARNING(
        "decoderFinishedEvent still has a value at the start of forwardAsync "
        "(microBatchId=%d). This indicates the previous decoderSync did not "
        "complete cleanly. Synchronizing and resetting to recover.",
        mMicroBatchId);
    decoderFinishedEvent->synchronize();
    decoderFinishedEvent.reset();
}
```

A single failed request can no longer cascade into permanent failure of the
micro-batch slot.

### 4. `generatedLength` safety clamp in `postProcessRequest()`

```cpp
auto const maxGeneratedLength
    = std::max<SizeType32>(0, static_cast<SizeType32>(maxSeqLength - llmReq.mPromptLen));
auto generatedLength = sequenceLengthsHostData[beam] - llmReq.mPromptLen;
if (generatedLength < 0 || generatedLength > maxGeneratedLength)
{
    TLLM_LOG_DEBUG(...);
    generatedLength = std::clamp(generatedLength, static_cast<SizeType32>(0), maxGeneratedLength);
}
```

Prevents `std::vector::assign` from throwing when `gatherTree` produces
invalid lengths for unconverged beams.

### 5. `setGeneratedTokens` updates `mLastTokens[beamId]`

So that the next decoder step sees a consistent "last token" for each beam
after gatherTree has replaced `mTokens` with the finalized beam paths.

## Benchmark

Workload: 42,021 requests, `concurrency=4`, `beam_width=128`, `max_tokens=8`.

| Variant                | QPS    | Latency P50 | Latency P99 | Failures |
|------------------------|--------|-------------|-------------|----------|
| Baseline (main)        | ~0     | hang        | hang        | all hang |
| **This PR**            | **27.54** | **0.138 s** | **0.235 s** | **0**    |

Note: Generation-logits reorder (the `reorderGenerationLogitsForBeamSearch` 
helper) was independently added to main while this PR was in review. The 
latest commit merges main and drops the duplicate; this PR now focuses 
only on the beam-search termination root fix.

## Files Changed

- `cpp/tensorrt_llm/batch_manager/trtGptModelInflightBatching.cpp`
  - `updateRequests()`: baseline snapshot, hard cutoff, kLENGTH broadcast,
    `needsGatheredBeamState` / `shouldFinalizeRequest`, `preDecodedTokens`
    gathered delta.
  - `forwardAsync()`: assertion → graceful recovery.
  - `postProcessRequest()`: clamp `generatedLength`, call generation-logits
    reorder.

- `cpp/tensorrt_llm/batch_manager/trtGptModelInflightBatching.h`
  - `reorderGenerationLogitsForBeamSearch` declaration, docstring update
    on `postProcessRequest`.
- `cpp/include/tensorrt_llm/batch_manager/llmRequest.h`
  - `setGeneratedTokens`: update `mLastTokens[beamId]` to the last gathered
    token.

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved beam search stability with enhanced error recovery mechanisms
  * Refined token sequence tracking for consistency during generation
  * Added safeguards to prevent request corruption during batch processing
  * Enhanced robustness for edge cases in sequence length handling

### CODE REVIEW

File: cpp/tensorrt_llm/batch_manager/trtGptModelInflightBatching.cpp [hunk 1]
```diff
@@ -1952,12 +2007,180 @@ void TrtGptModelInflightBatching::postProcessRequest(
...
+
+    if (!canReliablyReorder)
+    {
+        TLLM_LOG_WARNING(
+            "[postProcessRequest] request %lu: skipping beam-search generation logits reordering because "
+            "the gathered beam lineage could not be reconstructed reliably.",
+            llmReq.mRequestId);
+        TLLM_LOG_TRACE("%s stop", __PRETTY_FUNCTION__);
+        return;
+    }
```
[the remark this answers is not in the document]
Comment (Doloxetine) [reply cpp/tensorrt_llm/batch_manager/trtGptModelInflightBatching.cpp:2298]: N/A, this code is inherited from main after conflict resolution

**Comment by karljang** [comment]
@Doloxetine, thank you for your contribution!. the root-cause analysis here is excellent. The step-by-step breakdown of how beam reordering desynchronizes both the GPU-side and CPU-side termination paths, and how that cascades into the decoderFinishedEvent slot-poisoning, is a really clear and well-reasoned writeup. The benchmark makes the impact concrete.

That said, the entire fix lives in `TrtGptModelInflightBatching`, which is part of the legacy TensorRT (C++ engine) backend. This path is being phased out in favor of the PyTorch backend [link], and the PyTorch beam-search path doesn't go through these `updateRequests`/`forwardAsync`/`postProcessRequest` methods, so it isn't affected by this bug. Because we're no longer taking changes into the legacy executor, I'm going to close this PR as not planned.

This isn't a reflection on the quality of the work, it's solid. If you hit a similar large-beam_width termination issue on the PyTorch backend, please do open an issue or PR there; we'd very much welcome it. Thanks again for the careful investigation.
